File: src/main/scala/VoteForLife.scala
package ca.hyperreal.calabexample

import java.awt.Color
import Color._
import ca.hyperreal.color.HSL

import  ca.hyperreal.calab.{CAEngineConstructor, CAEngine, Universe}


class VoteForLife extends CAEngineConstructor
{
	val RULE = """(\d+)"""r
	
	def apply( rule: String ) =
	{
		if (RULE.pattern.matcher( rule ).matches)
		{
		val RULE(l) = rule
		
			Some( new VoteForLifeEngine(string(l)) )
		}
		else
			None
	}
	
	override def toString = "Vote"
}

class VoteForLifeEngine( lives: Set[Int] ) extends CAEngine
{
	def apply( x: Int, y: Int, u: Universe )
	{
	var neighbours = 0
	
		for (i <- -1 to 1; j <- -1 to 1)
			neighbours += u.read( x + i, y + j )

		u.write( x, y, if (lives( neighbours )) 1 else 0 )
	}
	
	val colors = Seq( DARK_GRAY.darker.darker, WHITE )
	
	val alive = 1
	
	override def toString = s"""Vote for Life [lives: {${lives.toList.sorted.mkString(",")}}]"""
}

File: src/main/scala/VoteForLife.java
package ca.hyperreal.calabexample;

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.HashSet;
import java.util.Set;

import ca.hyperreal.color.HSL;
import ca.hyperreal.calab.CAEngineConstructor;
import ca.hyperreal.calab.CAEngine;

import scala.*;


public class VoteForLife implements CAEngineConstructor
{
	final Pattern RULE = Pattern.compile( "(\\d+)" );
	final Option<CAEngine> None = Option.empty();
	
	public Option<CAEngine> instance( String rule )
	{
	Matcher m = RULE.matcher( rule );
		
		if (m.matches())
		{
		HashSet<Integer> lives = new HashSet<Integer>();
		
			for (int i = 0; i < rule.length(); i++)
				lives.add( rule.charAt(i) - '0' );
				
			return new Some( new VoteForLifeEngine(lives) );
		}
		else
			return None;
	}
	
	public String toString()
	{
		return "Vote";
	}
}